src/algebra/quadratic_discriminant.lean

```diff
@@ -0,0 +1,176 @@
+/-
+  Copyright (c) 2019 Zhouhang Zhou. All rights reserved.
+  Released under Apache 2.0 license as described in the file LICENSE.
+  Authors: Zhouhang Zhou
+  -/
+
+import algebra.ordered_field
+import tactic.linarith tactic.ring
+
+/-!
+  # Quadratic discriminants and roots of a quadratic
+
+  This file defines the discriminant of a quadratic and prove their
+
+  ## Main definition
+
+  The discriminant of a quadratic `a*x*x + b*x + c` is `b*b - 4*a*c`.
+
+  ## Main statements
+  • Roots of a quadratic can be written as `(-b + s) / (2 * a)` or `(-b - s) / (2 * a)`,
+    where `s` is the square root of the discriminant.
+  • If the discriminant has no square root, then the corresponding quadratic has no root.
+  • If a quadratic is always non-negative, then its discriminant is non-positive.
+
+  ## Tags
+
+  polynomial, quadratic, discriminant, root
+  -/
+
+variables {α : Type*}
+
+section lemmas
+
+variables [linear_ordered_field α] {a b c : α}
+
+lemma exists_le_mul : ∀ a : α, ∃ x : α, a ≤ x * x :=
+begin
+  assume a, cases le_total 1 a with ha ha,
+  { use a, exact le_mul_of_ge_one_left (by linarith) ha },
+  { use 1, linarith }
+end
+
+lemma exists_lt_mul : ∀ a : α, ∃ x : α, a < x * x :=
+begin
+  assume a, rcases (exists_le_mul a) with ⟨x, hx⟩,
+  cases le_total 0 x with hx' hx',
+  { use (x + 1),
+    have : (x+1)*(x+1) = x*x + 2*x + 1, ring,
+    exact lt_of_le_of_lt hx (by rw this; linarith) },
+  { use (x - 1),
+    have : (x-1)*(x-1) = x*x - 2*x + 1, ring,
+    exact lt_of_le_of_lt hx (by rw this; linarith) }
+end
+
+end lemmas
+
+variables [linear_ordered_field α] {a b c x : α}
+
+/-- Discriminant of a quadratic -/
+def discrim [ring α] (a b c : α) : α := b^2 - 4 * a * c
```

Comment:
Use a new type. You now have assumptions `linear_ordered_field` and `ring` on the same type `α`.